### PR TITLE
fix: Avoid `overlay` panic on valid Unicode input, Postgres compatibility

### LINKED ISSUE #22044: `overlay` panics on valid utf-8 input
### Describe the bug

```
  SELECT overlay('aébc' placing 'X' from 3 for 1);
-- Panics; correct result is 'aéXc'

  SELECT overlay('aébc' placing 'ZZ' from 2 for 2);
-- Returns 'aZZbc'; correct result is 'aZZc'
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### LINKED ISSUE #22045: `overlay` behavior inconsistent with Postgres / SQL spec
### Describe the bug

```
  SELECT overlay('abc' placing 'X' from 4 for 1);
  -- DF returns: 'X'
  -- Postgres returns: 'abcX'
```

```
  SELECT overlay('abc' placing 'X' from 0 for 1);
  -- DF returns: 'Xabc'
  -- Postgres produces "ERROR: negative substring length not allowed"
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by neilconway**
## Which issue does this PR close?

- Closes #22044.
- Closes #22045.

## Rationale for this change

`overlay` mixed up character-based and byte-based indexing, which resulted in either panicking or returning incorrect results on multi-byte Unicode input. While looking at this, I also fixed a few places where `overlay`'s behavior was inconsistent with Postgres and the SQL spec, refactored the code, and added a benchmark.

## What changes are included in this PR?

* Fix panic or wrong results on multibyte Unicode input
* Align edge-case behavior with Postgres: start past end appends replacement after the full input, `start <= 0` errors
* Refactored the code to reduce redundancy and clarify variable names
* Add SLT tests for Unicode input and behavior changes for Postgres compatibility
* Add benchmark, in anticipation of subsequent performance optimization work

## Are these changes tested?

Yes, and new tests added.

## Are there any user-facing changes?

Yes. The behavior of `overlay` has changed for multibyte Unicode inputs, as well as the edge cases enumerated above.

### CODE REVIEW

File: datafusion/functions/src/core/overlay.rs [hunk 1]
```diff
@@ -112,140 +112,170 @@ impl ScalarUDFImpl for OverlayFunc {
...
+            .zip($characters_array.iter())
+            .zip($pos_array.iter())
+            .zip($len_array.iter())
+            .map(|(((string, characters), start_pos), len)| {
+                match (string, characters, start_pos, len) {
+                    (Some(string), Some(characters), Some(start_pos), Some(len)) => {
+                        if start_pos < 1 {
+                            return exec_err!("negative substring length not allowed");
+                        }
+                        let string_char_len = string.chars().count() as i64;
```
Comment (kosiew) [inline datafusion/functions/src/core/overlay.rs]: Small optimization idea: the four-argument path calls `string.chars().count()` to clamp `len`, but `overlay_one` ends up computing the character length again internally. Since `overlay_one` already handles lengths that extend past the end of the string, it might be simpler to pass `len` through directly or move the clamp logic into `overlay_one` so we avoid the extra Unicode scan per row.

**Review by kosiew — COMMENTED** [review]
@neilconway 
Nice improvement overall. I just had one small performance suggestion around the four-argument overlay path.

**Review by kosiew — APPROVED** [review]

File: datafusion/functions/src/core/overlay.rs (hunk 1 above)
Comment (neilconway) [reply datafusion/functions/src/core/overlay.rs]: Good call! I was going to defer optimization work until a subsequent PR, but this is a straightforward cleanup. On closer inspection, we can just remove the clamp here, because `overlay_one` can handle length past the end of the string.

**Comment by neilconway** [comment]
@kosiew Is this okay to merge, do you think?

**Review by Dandandan — APPROVED** [review]
